npm/private/utils.bzl

```diff
@@ -151,36 +303,70 @@ def _convert_v6_packages(packages):
-        for key in ["dependencies", "optionalDependencies"]:
-            deps = package_info.get(key, None)
-            if deps != None:
-                dependencies = {}
-                for dep_name, dep_version in deps.items():
-                    dependencies[dep_name] = _convert_pnpm_v6_package_name(dep_version)
-                package_info[key] = dependencies
-
-        package_key = _convert_pnpm_v6_package_name(package)
+    for package_path, package_snapshot in packages.items():
+        if "resolution" not in package_snapshot:
+            msg = "package {} has no resolution field".format(package_path)
+            fail(msg)
+
+        package_path = _convert_pnpm_v6_v9_version_peer_dep(package_path)
+
+        if package_path.startswith("file:"):
+            # direct reference to file
+            package_key, name, version, friendly_version = _convert_v5_v6_file_package(package_path, package_snapshot)
+        elif "name" in package_snapshot and "version" in package_snapshot:
+            # key/path is complicated enough the real name+version are properties
+            name = package_snapshot["name"]
+            version = package_path
+            friendly_version = package_snapshot["version"]
+            package_key = package_path
+        elif package_path.startswith("/"):
+            # an aliased dependency
+            name, version = package_path[1:].rsplit("@", 1)
+
+            # TODO: dont strip twice, but 'friendly_version' may have already been converted from v6
+            friendly_version = _strip_v5_peer_dep_or_patched_version(_strip_v6_peer_dep_or_patched_version(version))
+            package_key = "{}@{}".format(name, version)
+        else:
+            msg = "unexpected package path: {}".format(package_path)
+            fail(msg)
+
+        package_info = _new_package_info(
+            id = package_snapshot.get("id", None),
+            name = name,
+            version = version,
+            friendly_version = friendly_version,
+            dependencies = _convert_pnpm_v6_v9_package_dependency_map(package_snapshot.get("dependencies", {})),
+            optional_dependencies = _convert_pnpm_v6_v9_package_dependency_map(package_snapshot.get("optionalDependencies", {})),
+            peer_dependencies = _convert_pnpm_v6_v9_package_dependency_map(package_snapshot.get("peerDependencies", {})),
+            dev = package_snapshot.get("dev", False),
+            has_bin = package_snapshot.get("hasBin", False),
+            optional = package_snapshot.get("optional", False),
+            requires_build = package_snapshot.get("requiresBuild", False),
+            resolution = package_snapshot.get("resolution"),
+        )
 
         if package_key in result:
             msg = "ERROR: duplicate package: {}\n\t{}\n\t{}".format(package_key, result[package_key], package_info)
-            fail(msg)
+
+            # buildifier: disable=print
+            print(msg)
```

Comment:
no longer a failure?

Reply:
I'll have to check again, but I think when upgrading the rules_js pnpm-lock it actually hit this case, there's a bug (which this PR neither introduces or fixes) where an oddly configured package has two entries with the same key.

Reply:
I'll have to check again, it's possible I switched it to `print` while debugging 🤔

Reply:
If you do a build on 2.x right now you'll see:
```
DEBUG: /Users/jason/dev/rules_js/npm/private/transitive_closure.bzl:164:18: ERROR: duplicate package info: lodash/4.17.21
	{"name": "lodash", "id": None, "version": "4.17.21", "friendly_version": "4.17.21", "resolution": {"integrity": "sha512-v2kDEe57lecTulaDIuNTPy3Ry4gLGJ6Z1O3vE1krgXZNrsQ+LFTGHVxVjcXPs17LhbZVGedAJv8XZ1tvj5FvSg=="}, "dependencies": {}, "optional_dependencies": {}, "dev": True, "optional": False, "has_bin": False, "requires_build": False}
	{"name": "lodash", "id": None, "version": "4.17.21", "friendly_version": "4.17.21", "resolution": {"integrity": "sha512-v2kDEe57lecTulaDIuNTPy3Ry4gLGJ6Z1O3vE1krgXZNrsQ+LFTGHVxVjcXPs17LhbZVGedAJv8XZ1tvj5FvSg==", "tarball": "file:npm/private/test/vendored/lodash-4.17.21.tgz"}, "dependencies": {}, "optional_dependencies": {}, "dev": True, "optional": False, "has_bin": False, "requires_build": False}
```

This PR now normalizes all the data in the parser, so that duplicate key is now caught here so needs to just be a warning like it was previously in `transitive_closure.bzl`. Basically the bug got pushed into the parser and I don't want this refactor to turn the warning into a failure. For lockfile v9 I hope to keep it as an error, it's just the v6 that needs to be non-breaking.